File: problem_0276/leetcode1.py
class Solution:
    def numWays(self, n: int, k: int) -> int:
        if n == 0: return 0
        # you can either use @functools.lru_cache
        # or you can implement an array to store
        # the values manually
        # more details: https://docs.python.org/3/library/functools.html
        @lru_cache(None)
        def total_ways(num):
            if num == 1:
                return k
            if num == 2:
                return k * k
            last = total_ways(num - 1)
            second_last = total_ways(num - 2)
            return (last + second_last) * (k - 1)

        return total_ways(n)

File: problem_0276/LeetCode1.java
class Solution {
    private int[] memo;

    public int numWays(int n, int k) {
        if (n == 0) return 0;
        memo = new int[n + 1];
        return totalWays(n, k);
    }

    private int totalWays(int posts, int k) {
        if (posts == 1) return k;
        if (posts == 2) return k * k;
        if (memo[posts - 1] == 0) {
            memo[posts - 1] = totalWays(posts - 1, k);
        }
        if (memo[posts - 2] == 0) {
            memo[posts - 2] = totalWays(posts - 2, k);
        }
        memo[posts] = (memo[posts - 1] + memo[posts - 2]) * (k - 1);
        return memo[posts];
    }
}
